index.bs

```diff
@@ -1450,6 +1451,60 @@ that are returned to the caller when a new credential is created, or a new asser
         [=extension identifier=] → [=client extension output=] entries produced by the extension's
         [=client extension processing=].
 
+    :   {{PublicKeyCredential/toJSON()}}
+    ::  This operation returns {{RegistrationResponseJSON}} or {{AuthenticationResponseJSON}}, that is a JSON-serialized
+        representation of the instance of {{PublicKeyCredential}} suitable for submission to a [=[RP]=] server as a DOMString
+        payload with `application/json` mimetype. The [=client=] is in charge of
+        [serializing values to JSON as usual](https://webidl.spec.whatwg.org/#idl-tojson-operation), but must take additional
+        steps to encode any {{ArrayBuffer}} values to {{DOMString}} values using [=base64url encoding=] to ensure compatibility
+        with JSON syntax. {{ArrayBuffer}} values to be encoded are identified according to the detected type of the instance's
+        `response` property.
+    ::  |clientExtensionResults| below should be the output of {{PublicKeyCredential/getClientExtensionResults()}}, with any
+        {{ArrayBuffer}} values encoded to {{DOMString}} values using [=base64url encoding=]. This may include {{ArrayBuffer}}
+        values from extensions registered in the IANA "WebAuthn Extension Identifiers" registry [[!IANA-WebAuthn-Registries]]
+        but not defined in [[#sctn-extensions]].
+</dl>
+
+<xmp class="idl">
+    typedef DOMString Base64URLString;
+    typedef (RegistrationResponseJSON or AuthenticationResponseJSON) PublicKeyCredentialJSON;
+
+    dictionary RegistrationResponseJSON {
+        Base64URLString id;
+        Base64URLString rawId;
+        AuthenticatorAttestationResponseJSON response;
+        DOMString?  authenticatorAttachment;
+        AuthenticationExtensionsClientOutputsJSON clientExtensionResults;
+        DOMString type;
+    };
+
+    dictionary AuthenticatorAttestationResponseJSON {
+        Base64URLString clientDataJSON;
+        Base64URLString attestationObject;
+        sequence<DOMString> transports;
+    };
+
+    dictionary AuthenticationResponseJSON {
```

Comment:
It's frustrating to me that `PublicKeyCredential` has been overloaded to mean two possible data structures, since `response` can be one of two types. I think this is a current developer pain point, that there aren't subclasses of `PublicKeyCredential` coming out of `.create()` and `.get()` to be uniquely typed. That's why I've been trying to emphasize the idea of "`RegistrationResponse`" and "`AuthenticationResponse`" here, to get to a 1:1 mapping between `navigator.credentials` method and return type.

To the suggestion that a `PublicKeyCredentialJSON` gets defined as a union type, is it a compromise to make it easier for the client to implement this feature? I personally want to see these additions err on the side of better _developer_ usability, and necessarily see this as hoisting some of this complexity onto the client. The aim is to make it _easier_ for devs to consume WebAuthn, and to achieve this I think we should be more straightforward/explicit with return types.

Reply:
> To the suggestion that a `PublicKeyCredentialJSON` gets defined as a union type, is it a compromise to make it easier for the client to implement this feature? I personally want to see these additions err on the side of better _developer_ usability, and necessarily see this as hoisting some of this complexity onto the client.

@kreichgauer It's been a few weeks, what do you currently think about my response above to your `PublicKeyCredentialJSON` suggestion?

Reply:
> I guess it depends on whether you think the hypothetical WebAuthn developer would first look at the PublicKeyCredential interface IDL or not? If they did, comparing the two and realizing that they're different in weird subtle ways (beyond the binary->base64url conversion) I think could be a bit puzzling? If on the other hand, we'd assume the hypothetical developer would only ever see the JSON IDL, I agree what you propose would be simpler to grok.

@kreichgauer Can we leave comments in IDL definitions? I want to believe a developer who doesn't want to get too deep into things will jump to these JSON methods, but it's probably more realistic that they'd stumble on `PublicKeyCredential` first and then the JSON IDL. If we could leave comments in the JSON IDL explaining why there are split types then perhaps we can have the best of both worlds.

Reply:
Yes, [IDL can contain comments](https://webidl.spec.whatwg.org/#prod-comment). The syntax is the same as in JavaScript.